docs/blueprints/selected-nodes-cache.md

```diff
@@ -0,0 +1,82 @@
+- satellite inspector, finding known offline nodes
+- repair, finding missing pieces from nodes that are offline or unreliable and also finding reliable nodes
+- garbage collection, reads node pieces counts
+- gracefulexit, getting exited nodes
+
+When selecting nodes to store files to, the following criteria must be met:
+- the node is not disqualified
+- the node is not suspended
+- the node has not exited
+- the node has sufficent free disk space
+- the node has been contacted recently
+- the node has participated in a sufficient number of audit
+- the nodes has sufficient uptime counts
+- it is a storage node type
+
+## Design
+
+We want to create a read-only cache that contains data from the nodes table that will be used to select storage nodes to upload files to.
+
+The cache should contain the following data:
+- already selected nodes and already selected **new** nodes (this may need to be 2 caches so we can select a smaller percentage of new nodes)
+- The only data that is sent to the uplink about the storage nodes is its address, typically being the IP and port. So the data that the cache should store is the storage node id, ip, port.
+
+The nodes table is the source of truth for the node data and should remain that way. The cache needs to be updated when it becomes stale. The cache should be updated when the following occur:
+
+A node should be removed from the cache when the following happen:
+- a node is disqualified, suspended, exited, out of space, hasn't been contacted recently
+
+A node should be updated in the cache when the following happen:
+- a node's address, port, or ip changes
+
+A node should be added to the cache when:
+- it is a new node on the network and is added to the nodes table for the first time (add to the unvetted selected nodes cache)
+
+A node should move from the unvetted to vetted cache when:
+- when it has completed enough audits and uptime checks to be vetted
+
+Using Redis:
+Using a remote cache like redis, might be preferred versus in memory since this cache will be used in each satellite api replica (which can currently scale up to 12).  The more api replicas we have, the more caches if we store in memory. 3 concerns about not using a remote cache are: 1) that this will cause more load on the database to keep all replica caches in sync (bandwidth and load), 2) every time we update the cache we need to lock it and potentially block other work, 3) every replica will use up this much more space in memory which will increase over time and add up with each new replica. We need to know the size this cache will be since our current redis instance handles up to 1 GB so we might need to increase the size. Currently about 80% of the nodes table qualifies to be selected for uploads, so the size of the cache for the largest cache should be about 8MB if we store the entire row in the cache, but if we only store the address and storagenode id should be less than a KB.
```

Comment:
Can you split the different points onto separate lines here, slightly easier to comment.

Also remote caches are going to be slower than in memory.

Reply:
PS: the node selection should take less than 0.5ms or better yet less than 0.1ms.